sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetPartitionReaderFactory.scala

```diff
@@ -0,0 +1,227 @@
+        // converted (`ParquetFilters.createFilter` returns an `Option`). That's why a `flatMap`
+        // is used here.
+        .flatMap(parquetFilters.createFilter)
+        .reduceOption(FilterApi.and)
+    } else {
+      None
+    }
+    // PARQUET_INT96_TIMESTAMP_CONVERSION says to apply timezone conversions to int96 timestamps'
+    // *only* if the file was created by something other than "parquet-mr", so check the actual
+    // writer here for this file.  We have to do this per-file, as each file in the table may
+    // have different writers.
+    // Define isCreatedByParquetMr as function to avoid unnecessary parquet footer reads.
+    def isCreatedByParquetMr: Boolean =
+      footerFileMetaData.getCreatedBy().startsWith("parquet-mr")
+
+    val convertTz =
+      if (timestampConversion && !isCreatedByParquetMr) {
+        Some(DateTimeUtils.getTimeZone(conf.get(SQLConf.SESSION_LOCAL_TIMEZONE.key)))
+      } else {
+        None
+      }
+
+    val attemptId = new TaskAttemptID(new TaskID(new JobID(), TaskType.MAP, 0), 0)
+    val hadoopAttemptContext = new TaskAttemptContextImpl(conf, attemptId)
+
+    // Try to push down filters when filter push-down is enabled.
+    // Notice: This push-down is RowGroups level, not individual records.
+    if (pushed.isDefined) {
+      ParquetInputFormat.setFilterPredicate(hadoopAttemptContext.getConfiguration, pushed.get)
+    }
+    val reader =
+      buildReaderFunc(split, file.partitionValues, hadoopAttemptContext, pushed, convertTz)
+    reader.initialize(split, hadoopAttemptContext)
+    reader
+  }
+
+  private def createRowBaseReader(file: PartitionedFile): RecordReader[Void, UnsafeRow] = {
+    buildReaderBase(file, createRowBaseReader0)
+  }
+
+  private def createRowBaseReader0(
+      split: ParquetInputSplit,
+      partitionValues: InternalRow,
+      hadoopAttemptContext: TaskAttemptContextImpl,
+      pushed: Option[FilterPredicate],
+      convertTz: Option[TimeZone]): RecordReader[Void, UnsafeRow] = {
+    logDebug(s"Falling back to parquet-mr")
+    val taskContext = Option(TaskContext.get())
+    // ParquetRecordReader returns UnsafeRow
+    val readSupport = new ParquetReadSupport(convertTz, enableVectorizedReader = false)
+    val reader = if (pushed.isDefined && enableRecordFilter) {
+      val parquetFilter = FilterCompat.get(pushed.get, null)
+      new ParquetRecordReader[UnsafeRow](readSupport, parquetFilter)
+    } else {
+      new ParquetRecordReader[UnsafeRow](readSupport)
+    }
+    val iter = new RecordReaderIterator(reader)
+    // SPARK-23457 Register a task completion lister before `initialization`.
+    taskContext.foreach(_.addTaskCompletionListener[Unit](_ => iter.close()))
+    reader
+  }
+
+  private def createVectorizedReader(file: PartitionedFile): VectorizedParquetRecordReader = {
+    val vectorizedReader =
+      buildReaderBase(file, createVectorizedReader0).asInstanceOf[VectorizedParquetRecordReader]
```

Comment:
@gengliangwang, why is this cast here? I expected it to be removed when the one in `createRowBaseReader` was removed.

Reply:
This is because here we need to call the method `initBatch` and `enableReturningBatches` of VectorizedParquetRecordReader. We can't just change the returned type as `RecordReader[Void, Object]` here.